### ISSUE TITLE
NaN loss with B0

### ISSUE THREAD

**Issue by digantamisra98** [lonePatient/EvoNorms_PyTorch#1]
My Implementation - https://github.com/digantamisra98/EvoNorm
My implementation of B0 is equivalent to yours, however, I am observing NaN loss in the training. Any pointers on what I might have done wrong would be great. 
Thanks!

**Comment by lonePatient** [comment]
@digantamisra98 if lr=0.1, I alse get NaN loss in the training (bn enb0 ens0). so I use lr=0.01

**Comment by digantamisra98** [comment]
@lonePatient I get NaN with lr = 0.001. Is my B0 implementation correct?

```
import torch
import torch.nn as nn

def instance_std(x, eps=1e-5):
    var = torch.var(x, dim = (2, 3), keepdim=True).expand_as(x)
    return torch.sqrt(var + eps)

def group_std(x, groups = 32, eps = 1e-5):
    N, C, H, W = x.size()
    x = torch.reshape(x, (N, groups, C // groups, H, W))
    var = torch.var(x, dim = (2, 3, 4), keepdim = True).expand_as(x)
    return torch.reshape(torch.sqrt(var + eps), (N, C, H, W))

class EvoNorm2D(nn.Module):

    def __init__(self, input, non_linear = True, version = 'S0', momentum = 0.9, training = True):
        super(EvoNorm2D, self).__init__()
        self.non_linear = non_linear
        self.version = version
        self.training = training
        self.momentum = momentum
        if self.version not in ['B0', 'S0']:
            raise ValueError("Invalid EvoNorm version")
        self.insize = input
        self.gamma = nn.Parameter(torch.ones(1, self.insize, 1, 1))
        self.beta = nn.Parameter(torch.zeros(1, self.insize, 1, 1))
        if self.non_linear:
            self.v = nn.Parameter(torch.ones(1,self.insize,1,1))
        self.register_buffer('running_var', torch.ones(1, self.insize, 1, 1))

        self.reset_parameters()

    def reset_parameters(self):
        self.running_var.fill_(1)
    
    def forward(self, x):
        if x.dim() != 4:
            raise ValueError('expected 4D input (got {}D input)'
                             .format(x.dim()))
        if self.version == 'S0':
            if self.non_linear:
                num = x * torch.sigmoid(self.v * x)
                return num / group_std(x) * self.gamma + self.beta
            else:
                return x * self.gamma + self.beta
        if self.version == 'B0':
            if self.training:
                var = torch.var(x, dim = (0, 2, 3), unbiased = False, keepdim = True).reshape(1, x.size(1), 1, 1)
                with torch.no_grad():
                   self.running_var.copy_(self.momentum * self.running_var + (1 - self.momentum) * var)
            else:
                var = self.running_var

            if self.non_linear:
                den = torch.max((var+self.eps).sqrt(), self.v * x + instance_std(x))
                return x / den * self.gamma + self.beta
            else:
                return x * self.gamma + self.beta
```

**Comment by sayakpaul** [comment]
Another thing that I noticed in @lonePatient's [S0 Implementation](https://github.com/lonePatient/EvoNorms_PyTorch/blob/master/models/normalization.py#L64) is the calculation of `num = torch.sigmoid(self.v * x)`. The original paper suggests the following though:

![image](https://user-images.githubusercontent.com/22957388/79629238-77cfab80-8165-11ea-85e2-81eb2bac7327.png)

**Comment by sayakpaul** [comment]
I am also facing `nan` loss with B0. However, I am trying this with a MiniInception module. Here's my [Colab Notebook](https://colab.research.google.com/drive/1aaA8owK4bBWvXn5d8YnGgWPiS0bfWhX8).

**Comment by digantamisra98** [comment]
Fixed in my new implementation.
